Correctly handle positional only arguments

**Describe the bug**
Currently, `docspec_python` does not create correct `Argument` objects because it adds an argument like:
```
{
 "name": "/",
"type": "POSITIONAL"
}
```

and does not use the arguement kind `Argument.Type.PositionalOnly`. 

**To Reproduce**
```python
@pytest.mark.skipif(sys.version_info < (3, 8), reason="requires python3.8 or higher")
@docspec_test()
def test_funcdef_7_posonly_args():
  """
  def func1(x, y=3, /, z=5, w=7): pass
  def func2(x, /, *v, a=1, b=2): pass
  def func3(x, /, *, a=1, b=2, **kwargs): pass
  """

  return [
    mkfunc('func1', None, [
      Argument('x', Argument.Type.PositionalOnly),
      Argument('y', Argument.Type.PositionalOnly, default_value='3'),
      Argument('z', Argument.Type.Positional, default_value='5'),
      Argument('w', Argument.Type.Positional, default_value='7'),
    ]),
    mkfunc('func2', None, [
      Argument('x', Argument.Type.PositionalOnly),
      Argument('v', Argument.Type.PositionalRemainder),
      Argument('a', Argument.Type.KeywordOnly, default_value='1'),
      Argument('b', Argument.Type.KeywordOnly, default_value='2'),
    ]),
    mkfunc('func3', None, [
      Argument('x', Argument.Type.PositionalOnly),
      Argument('a', Argument.Type.KeywordOnly, default_value='1'),
      Argument('b', Argument.Type.KeywordOnly, default_value='2'),
      Argument('kwargs', Argument.Type.KeywordRemainder),
    ]),
  ]
```

Then run tests and see the diff: 

```
E         {
E           "name": "funcdef_7_posonly_args",
E           "location": null,
E           "docstring": null,
E           "members": [
E             {
E               "name": "func1",
E               "location": null,
E               "docstring": null,
E               "modifiers": null,
E               "args": [
E                 {
E                   "name": "x",
E       -           "type": "POSITIONAL"
E       +           "type": "POSITIONAL_ONLY"
E       ?                              +++++
E                 },
E                 {
E                   "name": "y",
E       -           "type": "POSITIONAL",
E       +           "type": "POSITIONAL_ONLY",
E       ?                              +++++
E                   "default_value": "3"
E       -         },
E       -         {
E       -           "name": "/",
E       -           "type": "POSITIONAL"
E                 },
E                 {
E                   "name": "z",
E                   "type": "POSITIONAL",
E                   "default_value": "5"
E                 },
E                 {
E                   "name": "w",
E                   "type": "POSITIONAL",
E                   "default_value": "7"
E                 }
E               ],
E               "return_type": null,
E               "decorations": [],
E               "type": "function"
E             },
E             {
E               "name": "func2",
E               "location": null,
E               "docstring": null,
E               "modifiers": null,
E               "args": [
E                 {
E                   "name": "x",
E       -           "type": "POSITIONAL"
E       +           "type": "POSITIONAL_ONLY"
E       ?                              +++++
E       -         },
E       -         {
E       -           "name": "/",
E       -           "type": "POSITIONAL"
E                 },
E                 {
E                   "name": "v",
E                   "type": "POSITIONAL_REMAINDER"
E                 },
E                 {
E                   "name": "a",
E                   "type": "KEYWORD_ONLY",
E                   "default_value": "1"
E                 },
E                 {
E                   "name": "b",
E                   "type": "KEYWORD_ONLY",
E                   "default_value": "2"
E                 }
E               ],
E               "return_type": null,
E               "decorations": [],
E               "type": "function"
E             },
E             {
E               "name": "func3",
E               "location": null,
E               "docstring": null,
E               "modifiers": null,
E               "args": [
E                 {
E                   "name": "x",
E       -           "type": "POSITIONAL"
E       +           "type": "POSITIONAL_ONLY"
E       ?                              +++++
E       -         },
E       -         {
E       -           "name": "/",
E       -           "type": "POSITIONAL"
E                 },
E                 {
E                   "name": "a",
E                   "type": "KEYWORD_ONLY",
E                   "default_value": "1"
E                 },
E                 {
E                   "name": "b",
E                   "type": "KEYWORD_ONLY",
E                   "default_value": "2"
E                 },
E                 {
E                   "name": "kwargs",
E                   "type": "KEYWORD_REMAINDER"
E                 }
E               ],
E               "return_type": null,
E               "decorations": [],
E               "type": "function"
E             }
E           ]
E         }
```

**Expected behavior**
The expected behaviour would be to produce `Argument` with kind `POSITIONAL_ONLY` and do not have the extra `/` argument in the list.

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -24,13 +24,14 @@ jobs:
           /opt/pipx
           /opt/pipx_bin
           /opt/hostedtoolcache/pyenv_root
-        key: *pyenv-config
+        # NOTE: Same as pyenv config
+        key: 3.10.2 pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
 
     - uses: gabrielfalcao/pyenv-action@v9
       if: steps.cache.outputs.cache-hit != 'true'
       with:
         default: 3.10.2
-        command: &pyenv-config pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
+        command: pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
 
     - name: Test tooling
       run: |

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -16,7 +16,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ['3.7.12'] #, '3.8.12', '3.9.10', '3.10.2']
+        python-version: ['3.7.12', '3.8.12', '3.9.10', '3.10.2']
         python-tooling-version: ['3.10.2']
         project: ['docspec', 'docspec-python']
     steps:

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -11,6 +11,27 @@ on:
     branches: [ develop ]
 
 jobs:
+
+  install-tooling:
+    runs-on: ubuntu-latest
+    steps:
+
+    - uses: gabrielfalcao/pyenv-action@v9
+      with:
+        default: 3.10.2
+        command: pip install -U pip
+
+    - name: Install tooling
+      run: pyenv local 3.10.2 && pyenv exec pip install pipx && pipx install poetry==1.2.0a2 slam-cli==1.0.0a9
+
+    - name: Cache tooling environment
+      uses: actions/upload-artifact@v2
+      with:
+        name: tooling
+        path: |
+          ~/.local
+          ~/.pyenv
+
   test:
     runs-on: ubuntu-latest
     strategy:
@@ -18,16 +39,25 @@ jobs:
         python-version: ['3.7.12', '3.8.12', '3.9.10', '3.10.2']
         project: ['docspec', 'docspec-python']
     steps:
+
+    - name: Materialize tooling environment
+      uses: actions/download-artifact@v2
+      with:
+        name: tooling
+
+    - name: Test tooling environment
+      run: |
+        poetry --version
+        slam --version
+
     - uses: actions/checkout@v2
+
     - name: Set up Python ${{ matrix.python-version }}
       uses: gabrielfalcao/pyenv-action@v9
       with:
         default: "${{ matrix.python-version }}"
         versions: 3.10.2
         command: pip install -U pip
-
-    - name: Install tooling
-      run: pyenv local 3.10.2 && pyenv exec pip install poetry==1.2.0a2 slam-cli==1.0.0a9
 
     - name: Install ${{ matrix.project }}
       run: |

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -41,7 +41,7 @@ jobs:
       run: slam install --only ${{ matrix.project }} --no-venv-check
 
     - name: Slam test
-      run: slam test
+      run: slam test ${{ matrix.project }}
 
   update-pr-numbers:
     if: github.event_name == 'pull_request'

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -15,26 +15,29 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        #python-version: ['3.7', '3.8', '3.9', '3.10', '3.x']
-        python-version: ['3.10', '3.x']
+        python-version: ['3.7', '3.8', '3.9', '3.10']
+        project: ['docspec', 'docspec-python']
     steps:
     - uses: actions/checkout@v2
     - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v2
+      uses: gabrielfalcao/pyenv-action@v9
       with:
-        python-version: ${{ matrix.python-version }}
-    - name: Install dependencies
+        default: "${{ matrix.python-version }}"code dot
+        versions: [ "3.10" ]
+        command: pip install -U pip
+
+    - name: Install tooling
+      run: pyenv local 3.10 && pyenv exec pip install poetry==1.2.0a2 slam-cli==1.0.0a8
+
+    - name: Install ${{ matrix.project }}
       run: |
-        python -m pip install \
-          poetry==1.2.0a2 \
-          slam-cli==1.0.0a8 \
-          mypy \
-          pytest \
-          types-deprecated \
-          types-termcolor \
-          ./docspec \
-          ./docspec-python
-    - name: Run tests
+        pyenv local "${{ matrix.python-version }}"
+        which pip
+        pyenv which pip
+        pip install ./docspec
+        pushd "${{ matrix.project }}" && poetry install --dev-only && popd
+
+    - name: Slam test
       run: slam test
 
   update-pr-numbers:

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -22,7 +22,7 @@ jobs:
         command: pip install -U pip
 
     - name: Install tooling
-      run: pyenv local 3.10.2 && pyenv exec pip install pipx && pipx install poetry==1.2.0a2 slam-cli==1.0.0a9
+      run: pyenv local 3.10.2 && pyenv exec pip install pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
 
     - name: Cache tooling environment
       uses: actions/upload-artifact@v2
@@ -33,6 +33,7 @@ jobs:
           ~/.pyenv
 
   test:
+    needs: install-tooling
     runs-on: ubuntu-latest
     strategy:
       matrix:

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -33,6 +33,7 @@ jobs:
         pip install -U pip pipx -q
         pipx install slam-cli==1.0.0a13
         pyenv local ${{ matrix.python-version }}
+        pip install -U pip -q
 
     - uses: actions/checkout@v2
 

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -104,7 +104,7 @@ jobs:
         which pip
         pyenv which pip
         pip install ./docspec
-        pushd "${{ matrix.project }}" && poetry install --dev-only && popd
+        pushd "${{ matrix.project }}" && poetry install --only=dev && popd
 
     - name: Slam test
       run: slam test

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -22,7 +22,7 @@ jobs:
     - name: Set up Python ${{ matrix.python-version }}
       uses: gabrielfalcao/pyenv-action@v9
       with:
-        default: "${{ matrix.python-version }}"code dot
+        default: "${{ matrix.python-version }}"
         versions: [ "3.10" ]
         command: pip install -U pip
 

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -12,74 +12,34 @@ on:
 
 jobs:
 
-  install-tooling:
-    runs-on: ubuntu-latest
-    steps:
-
-    - name: Cache tooling environment
-      id: cache
-      uses: actions/cache@v2
-      with:
-        path: |
-          /opt/pipx_bin
-          /opt/pipx
-          /opt/hostedtoolcache/pyenv_root
-        # NOTE: Same as pyenv config
-        key: 3.10.2 pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
-
-    - uses: gabrielfalcao/pyenv-action@v9
-      if: steps.cache.outputs.cache-hit != 'true'
-      with:
-        default: 3.10.2
-        command: pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
-
-    - name: Test tooling
-      run: |
-        poetry --version
-        slam --version
-
-        pyenv which pip
-        pyenv local 3.10.2
-        pyenv which pip
-
-        du -hs /opt/pipx_bin /opt/pipx /opt/hostedtoolcache/pyenv_root
-
-    - name: Cache tooling environment
-      uses: actions/upload-artifact@v2
-      with:
-        name: tooling
-        path: |
-          /opt/pipx_bin
-          /opt/pipx
-          /opt/hostedtoolcache/pyenv_root
-
   test:
-    needs: install-tooling
     runs-on: ubuntu-latest
     strategy:
       matrix:
+        python-tooling-version: ['3.10.2']
         python-version: ['3.7.12', '3.8.12', '3.9.10', '3.10.2']
         project: ['docspec', 'docspec-python']
     steps:
 
-    - name: Materialize tooling environment
-      uses: actions/download-artifact@v2
+    - uses: gabrielfalcao/pyenv-action@v9
+      if: steps.cache.outputs.cache-hit != 'true'
       with:
-        name: tooling
+        default: "${{ matrix.python-version }}"
+        versions: "${{ matrix.python-tooling-version }}"
+        command: pip install -U pip
 
-    - name: Test tooling environment
+    - name: Install tooling
       run: |
+        which pip
+        pyenv local ${{ matrix.python-tooling-version }}
+        which pip
+        pip install pipx -q
+        pipx install poetry
+        pipx install slam-cli==1.0.0a9
         poetry --version
         slam --version
 
     - uses: actions/checkout@v2
-
-    - name: Set up Python ${{ matrix.python-version }}
-      uses: gabrielfalcao/pyenv-action@v9
-      with:
-        default: "${{ matrix.python-version }}"
-        versions: 3.10.2
-        command: pip install -U pip
 
     - name: Install ${{ matrix.project }}
       run: |

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -37,26 +37,12 @@ jobs:
       run: |
         poetry --version
         slam --version
-        which poetry
-        which slam
-        which pipx
-        which python
+
         pyenv which pip
         pyenv local 3.10.2
         pyenv which pip
-        pyenv which python
 
-        echo
-        echo ls ~/
-        ls ~/
-
-        echo
-        echo ls /opt
-        ls /opt
-
-        echo
-        echo ls /opt/hostedtoolcache
-        ls /opt/hostedtoolcache
+        du -hs /opt/pipx_bin /opt/pipx /opt/hostedtoolcache/pyenv_root
 
     - name: Cache tooling environment
       uses: actions/upload-artifact@v2

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -52,11 +52,11 @@ jobs:
 
         echo
         echo ls /opt
-        ls ~/opt
+        ls /opt
 
         echo
         echo ls /hostedtoolcache
-        ls ~/hostedtoolcache
+        ls /hostedtoolcache
 
     - name: Cache tooling environment
       uses: actions/upload-artifact@v2

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -30,9 +30,8 @@ jobs:
       run: |
         pyenv install ${{ matrix.python-tooling-version }} -s
         pyenv local ${{ matrix.python-tooling-version }}
-        pip install pipx -q
-        #pipx install slam-cli==1.0.0a12
-        pipx install git+https://github.com/NiklasRosenstein/slam.git
+        pip install -U pip pipx -q
+        pipx install slam-cli==1.0.0a13 -q
         pyenv local ${{ matrix.python-version }}
 
     - uses: actions/checkout@v2
@@ -49,13 +48,16 @@ jobs:
       contents: write
     runs-on: ubuntu-latest
     steps:
+
     - uses: actions/checkout@v2
+
     - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@v2
-      with:
-        python-version: '3.10'
+      with: { python-version: '3.10' }
+
     - name: Install Slam
-      run: pip install slam-cli==1.0.0a12
+      run: pip install slam-cli==1.0.0a13
+
     - name: Update PR references in changelogs
       run: slam -vv changelog update-pr --use github-actions
 

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -23,7 +23,7 @@ jobs:
       uses: gabrielfalcao/pyenv-action@v9
       with:
         default: "${{ matrix.python-version }}"
-        versions: [ "3.10" ]
+        versions: 3.10.0
         command: pip install -U pip
 
     - name: Install tooling

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -15,6 +15,7 @@ jobs:
   test:
     runs-on: ubuntu-latest
     strategy:
+      fail-fast: false
       matrix:
         python-version: ['3.7.12', '3.8.12', '3.9.10', '3.10.2']
         python-tooling-version: ['3.10.2']

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -101,7 +101,7 @@ jobs:
         which pip
         pyenv which pip
         pip install ./docspec
-        pushd "${{ matrix.project }}" && poetry install --only=dev && popd
+        pushd "${{ matrix.project }}" && poetry install --no-root --E dev && popd
 
     - name: Slam test
       run: slam test

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -28,9 +28,9 @@ jobs:
 
     - uses: gabrielfalcao/pyenv-action@v9
       if: steps.cache.outputs.cache-hit != 'true'
-      with: &pyenv-config
+      with:
         default: 3.10.2
-        command: pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
+        command: &pyenv-config pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
 
     - name: Test tooling
       run: |

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -31,7 +31,7 @@ jobs:
         pyenv install ${{ matrix.python-tooling-version }} -s
         pyenv local ${{ matrix.python-tooling-version }}
         pip install -U pip pipx -q
-        pipx install slam-cli==1.0.0a13 -q
+        pipx install slam-cli==1.0.0a13
         pyenv local ${{ matrix.python-version }}
 
     - uses: actions/checkout@v2

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -27,7 +27,7 @@ jobs:
         command: pip install -U pip
 
     - name: Install tooling
-      run: pyenv local 3.10 && pyenv exec pip install poetry==1.2.0a2 slam-cli==1.0.0a8
+      run: pyenv local 3.10 && pyenv exec pip install poetry==1.2.0a2 slam-cli==1.0.0a9
 
     - name: Install ${{ matrix.project }}
       run: |
@@ -52,7 +52,7 @@ jobs:
       with:
         python-version: '3.10'
     - name: Install Slam
-      run: pip install slam-cli==1.0.0a8
+      run: pip install slam-cli==1.0.0a9
     - name: Update PR references in changelogs
       run: slam -vv changelog update-pr --use github-actions
 

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -23,11 +23,11 @@ jobs:
       uses: gabrielfalcao/pyenv-action@v9
       with:
         default: "${{ matrix.python-version }}"
-        versions: 3.10.0
+        versions: 3.10.2
         command: pip install -U pip
 
     - name: Install tooling
-      run: pyenv local 3.10 && pyenv exec pip install poetry==1.2.0a2 slam-cli==1.0.0a9
+      run: pyenv local 3.10.2 && pyenv exec pip install poetry==1.2.0a2 slam-cli==1.0.0a9
 
     - name: Install ${{ matrix.project }}
       run: |

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -55,8 +55,8 @@ jobs:
         ls /opt
 
         echo
-        echo ls /hostedtoolcache
-        ls /hostedtoolcache
+        echo ls /opt/hostedtoolcache
+        ls /opt/hostedtoolcache
 
     - name: Cache tooling environment
       uses: actions/upload-artifact@v2

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -50,10 +50,6 @@ jobs:
         ls ~/
 
         echo
-        echo ls ~/.local
-        ls ~/.local
-
-        echo
         echo ls /opt
         ls ~/opt
 

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -15,7 +15,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ['3.7', '3.8', '3.9', '3.10']
+        python-version: ['3.7.12', '3.8.12', '3.9.10', '3.10.2']
         project: ['docspec', 'docspec-python']
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -24,11 +24,11 @@ jobs:
           /opt/pipx
           /opt/pipx_bin
           /opt/hostedtoolcache/pyenv_root
-        key: cache-tooling
+        key: *pyenv-config
 
     - uses: gabrielfalcao/pyenv-action@v9
       if: steps.cache.outputs.cache-hit != 'true'
-      with:
+      with: &pyenv-config
         default: 3.10.2
         command: pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
 

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -21,8 +21,8 @@ jobs:
       uses: actions/cache@v2
       with:
         path: |
-          /opt/pipx
           /opt/pipx_bin
+          /opt/pipx
           /opt/hostedtoolcache/pyenv_root
         # NOTE: Same as pyenv config
         key: 3.10.2 pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
@@ -65,7 +65,7 @@ jobs:
         path: |
           /opt/pipx_bin
           /opt/pipx
-          /hostedtoolcache/pyenv_root
+          /opt/hostedtoolcache/pyenv_root
 
   test:
     needs: install-tooling

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -22,7 +22,18 @@ jobs:
         command: pip install -U pip
 
     - name: Install tooling
-      run: pyenv local 3.10.2 && pyenv exec pip install pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
+      run: pyenv local 3.10.2 && pyenv exec pip install pipx && pipx install poetry -q && pipx install slam-cli==1.0.0a9 -q
+
+    - name: Test tooling
+      run: |
+        poetry --version
+        slam --version
+        which poetry
+        which slam
+        which pipx
+        penv which pip
+        penv local 3.10.2
+        penv which pip
 
     - name: Cache tooling environment
       uses: actions/upload-artifact@v2

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -39,7 +39,7 @@ jobs:
     - uses: actions/checkout@v2
 
     - name: Install ${{ matrix.project }}
-      run: slam install --only ${{ matrix.project }} --no-venv-check
+      run: slam install --only ${{ matrix.project }} --no-venv-check -v
 
     - name: Slam test
       run: slam test ${{ matrix.project }}

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -32,7 +32,7 @@ jobs:
         pyenv install ${{ matrix.python-tooling-version }} -s
         pyenv local ${{ matrix.python-tooling-version }}
         pip install -U pip pipx -q
-        pipx install slam-cli==1.0.0a13
+        pipx install slam-cli==1.0.0a14
         pyenv local ${{ matrix.python-version }}
         pip install -U pip -q
 
@@ -58,7 +58,7 @@ jobs:
       with: { python-version: '3.10' }
 
     - name: Install Slam
-      run: pip install slam-cli==1.0.0a13
+      run: pip install slam-cli==1.0.0a14
 
     - name: Update PR references in changelogs
       run: slam -vv changelog update-pr --use github-actions

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -16,7 +16,18 @@ jobs:
     runs-on: ubuntu-latest
     steps:
 
+    - name: Cache tooling environment
+      id: cache
+      uses: actions/cache@v2
+      with:
+        path: |
+          /opt/pipx
+          /opt/pipx_bin
+          /opt/hostedtoolcache/pyenv_root
+        key: cache-tooling
+
     - uses: gabrielfalcao/pyenv-action@v9
+      if: steps.cache.outputs.cache-hit != 'true'
       with:
         default: 3.10.2
         command: pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
@@ -29,10 +40,10 @@ jobs:
         which slam
         which pipx
         which python
-        penv which pip
-        penv local 3.10.2
-        penv which pip
-        penv which python
+        pyenv which pip
+        pyenv local 3.10.2
+        pyenv which pip
+        pyenv which python
 
         echo
         echo ls ~/
@@ -56,6 +67,7 @@ jobs:
         name: tooling
         path: |
           /opt/pipx_bin
+          /opt/pipx
           /hostedtoolcache/pyenv_root
 
   test:

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -16,26 +16,25 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-tooling-version: ['3.10.2']
         python-version: ['3.7.12', '3.8.12', '3.9.10', '3.10.2']
+        python-tooling-version: ['3.10.2']
         project: ['docspec', 'docspec-python']
     steps:
 
     - uses: gabrielfalcao/pyenv-action@v9
-      if: steps.cache.outputs.cache-hit != 'true'
       with:
         default: "${{ matrix.python-version }}"
-        versions: "${{ matrix.python-tooling-version }}"
         command: pip install -U pip
 
     - name: Install tooling
       run: |
+        pyenv install ${{ matrix.python-tooling-version }} -s
         which pip
         pyenv local ${{ matrix.python-tooling-version }}
         which pip
         pip install pipx -q
         pipx install poetry
-        pipx install slam-cli==1.0.0a9
+        pipx install slam-cli==1.0.0a12
         poetry --version
         slam --version
 
@@ -64,7 +63,7 @@ jobs:
       with:
         python-version: '3.10'
     - name: Install Slam
-      run: pip install slam-cli==1.0.0a9
+      run: pip install slam-cli==1.0.0a12
     - name: Update PR references in changelogs
       run: slam -vv changelog update-pr --use github-actions
 

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -22,7 +22,7 @@ jobs:
         command: pip install -U pip
 
     - name: Install tooling
-      run: pyenv local 3.10.2 && pyenv exec pip install pipx && pipx install poetry -q && pipx install slam-cli==1.0.0a9 -q
+      run: pyenv local 3.10.2 && pyenv exec pip install pipx -q && pipx install poetry && pipx install slam-cli==1.0.0a9
 
     - name: Test tooling
       run: |

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -29,24 +29,16 @@ jobs:
     - name: Install tooling
       run: |
         pyenv install ${{ matrix.python-tooling-version }} -s
-        which pip
         pyenv local ${{ matrix.python-tooling-version }}
-        which pip
         pip install pipx -q
-        pipx install poetry
-        pipx install slam-cli==1.0.0a12
-        poetry --version
-        slam --version
+        #pipx install slam-cli==1.0.0a12
+        pipx install git+https://github.com/NiklasRosenstein/slam.git
+        pyenv local ${{ matrix.python-version }}
 
     - uses: actions/checkout@v2
 
     - name: Install ${{ matrix.project }}
-      run: |
-        pyenv local "${{ matrix.python-version }}"
-        which pip
-        pyenv which pip
-        pip install ./docspec
-        pushd "${{ matrix.project }}" && poetry install --no-root --E dev && popd
+      run: slam install --only ${{ matrix.project }} --no-venv-check
 
     - name: Slam test
       run: slam test

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -16,7 +16,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ['3.7.12', '3.8.12', '3.9.10', '3.10.2']
+        python-version: ['3.7.12'] #, '3.8.12', '3.9.10', '3.10.2']
         python-tooling-version: ['3.10.2']
         project: ['docspec', 'docspec-python']
     steps:

--- a/.github/workflows/python.yml
+++ b/.github/workflows/python.yml
@@ -19,10 +19,7 @@ jobs:
     - uses: gabrielfalcao/pyenv-action@v9
       with:
         default: 3.10.2
-        command: pip install -U pip
-
-    - name: Install tooling
-      run: pyenv local 3.10.2 && pyenv exec pip install pipx -q && pipx install poetry && pipx install slam-cli==1.0.0a9
+        command: pip install -U pip && pip install pipx -q && which pip && which pipx && pipx install poetry && pipx install slam-cli==1.0.0a9
 
     - name: Test tooling
       run: |
@@ -31,17 +28,35 @@ jobs:
         which poetry
         which slam
         which pipx
+        which python
         penv which pip
         penv local 3.10.2
         penv which pip
+        penv which python
+
+        echo
+        echo ls ~/
+        ls ~/
+
+        echo
+        echo ls ~/.local
+        ls ~/.local
+
+        echo
+        echo ls /opt
+        ls ~/opt
+
+        echo
+        echo ls /hostedtoolcache
+        ls ~/hostedtoolcache
 
     - name: Cache tooling environment
       uses: actions/upload-artifact@v2
       with:
         name: tooling
         path: |
-          ~/.local
-          ~/.pyenv
+          /opt/pipx_bin
+          /hostedtoolcache/pyenv_root
 
   test:
     needs: install-tooling

--- a/.gitignore
+++ b/.gitignore
@@ -7,6 +7,7 @@
 *.egg
 build/
 .vscode/
+.python-version
 
 # Generate files for documentation.
 changelog.md

--- a/docspec-python/.changelog/_unreleased.toml
+++ b/docspec-python/.changelog/_unreleased.toml
@@ -16,6 +16,7 @@ authors = [
     "@NiklasRosenstein",
     "@tristanlatr",
 ]
+pr = "https://github.com/NiklasRosenstein/docspec/pulls/58"
 issues = [
-    "https://github.com/NiklasRosenstein/docspec/issues/57"
+    "https://github.com/NiklasRosenstein/docspec/issues/57",
 ]

--- a/docspec-python/.changelog/_unreleased.toml
+++ b/docspec-python/.changelog/_unreleased.toml
@@ -26,12 +26,14 @@ id = "baf461c7-737f-4cda-b2fd-1213b7637281"
 type = "refactor"
 description = "use `nr.util.scanner.Scanner` instead of homebrew `ListScanner` class"
 author = "@NiklasRosenstein"
+pr = "https://github.com/NiklasRosenstein/docspec/pulls/58"
 
 [[entries]]
 id = "0338523c-6ec7-42f0-aba6-7cf9266302f0"
 type = "fix"
 description = "fix parsing of positional only arguments"
 author = "@NiklasRosenstein"
+pr = "https://github.com/NiklasRosenstein/docspec/pulls/58"
 issues = [
     "https://github.com/NiklasRosenstein/docspec/issues/57",
 ]

--- a/docspec-python/.changelog/_unreleased.toml
+++ b/docspec-python/.changelog/_unreleased.toml
@@ -20,3 +20,18 @@ pr = "https://github.com/NiklasRosenstein/docspec/pulls/58"
 issues = [
     "https://github.com/NiklasRosenstein/docspec/issues/57",
 ]
+
+[[entries]]
+id = "baf461c7-737f-4cda-b2fd-1213b7637281"
+type = "refactor"
+description = "use `nr.util.scanner.Scanner` instead of homebrew `ListScanner` class"
+author = "@NiklasRosenstein"
+
+[[entries]]
+id = "0338523c-6ec7-42f0-aba6-7cf9266302f0"
+type = "fix"
+description = "fix parsing of positional only arguments"
+author = "@NiklasRosenstein"
+issues = [
+    "https://github.com/NiklasRosenstein/docspec/issues/57",
+]

--- a/docspec-python/.changelog/_unreleased.toml
+++ b/docspec-python/.changelog/_unreleased.toml
@@ -7,3 +7,15 @@ pr = "https://github.com/NiklasRosenstein/docspec/pulls/55"
 issues = [
     "https://github.com/NiklasRosenstein/docspec/issues/53",
 ]
+
+[[entries]]
+id = "2aa754a8-829e-40b0-947d-ebcb1d72284d"
+type = "tests"
+description = "add `test_funcdef_7_posonly_args` unit test to test various more combinations of function arguments, including `POSITIONAL_ONLY`)"
+authors = [
+    "@NiklasRosenstein",
+    "@tristanlatr",
+]
+issues = [
+    "https://github.com/NiklasRosenstein/docspec/issues/57"
+]

--- a/docspec-python/.changelog/_unreleased.toml
+++ b/docspec-python/.changelog/_unreleased.toml
@@ -16,7 +16,6 @@ authors = [
     "@NiklasRosenstein",
     "@tristanlatr",
 ]
-pr = "https://github.com/NiklasRosenstein/docspec/pulls/58"
 issues = [
     "https://github.com/NiklasRosenstein/docspec/issues/57",
 ]

--- a/docspec-python/.changelog/_unreleased.toml
+++ b/docspec-python/.changelog/_unreleased.toml
@@ -16,6 +16,7 @@ authors = [
     "@NiklasRosenstein",
     "@tristanlatr",
 ]
+pr = "https://github.com/NiklasRosenstein/docspec/pulls/58"
 issues = [
     "https://github.com/NiklasRosenstein/docspec/issues/57",
 ]

--- a/docspec-python/pyproject.toml
+++ b/docspec-python/pyproject.toml
@@ -14,9 +14,9 @@ packages = [{ include = "docspec_python", from="src" }]
   "nr.util" = "^0.6.6"
 
   [tool.poetry.dev-dependencies]
-  types-Deprecated = "*"
-  pytest = "*"
   mypy = "*"
+  pytest = "*"
+  types-Deprecated = "*"
 
   [tool.poetry.scripts]
   docspec-python = "docspec_python.__main__:main"

--- a/docspec-python/pyproject.toml
+++ b/docspec-python/pyproject.toml
@@ -11,6 +11,7 @@ packages = [{ include = "docspec_python", from="src" }]
   [tool.poetry.dependencies]
   python = "^3.7"
   docspec = "^1.2.0"
+  "nr.util" = "^0.6.4"
 
   [tool.poetry.dev-dependencies]
   types-Deprecated = "*"

--- a/docspec-python/pyproject.toml
+++ b/docspec-python/pyproject.toml
@@ -11,7 +11,7 @@ packages = [{ include = "docspec_python", from="src" }]
   [tool.poetry.dependencies]
   python = "^3.7"
   docspec = "^1.2.0"
-  "nr.util" = "^0.6.4"
+  "nr.util" = "^0.6.6"
 
   [tool.poetry.dev-dependencies]
   types-Deprecated = "*"
@@ -19,7 +19,6 @@ packages = [{ include = "docspec_python", from="src" }]
   mypy = "*"
 
   [tool.poetry.scripts]
-
   docspec-python = "docspec_python.__main__:main"
 
 [tool.slam]

--- a/docspec-python/src/docspec_python/parser.py
+++ b/docspec-python/src/docspec_python/parser.py
@@ -271,7 +271,7 @@ class Parser:
       return_type=return_,
       decorations=decorations)
 
-  def parse_argument(self, node: t.Union[Leaf, Node], argtype: Argument.Type, scanner: 'Scanner[Leaf | Node]') -> Argument:
+  def parse_argument(self, node: t.Union[Leaf, Node, None], argtype: Argument.Type, scanner: 'Scanner[Leaf | Node]') -> Argument:
     """
     Parses an argument from the AST. *node* must be the current node at
     the current position of the *scanner*. The scanner is used to extract
@@ -293,6 +293,7 @@ class Parser:
         raise RuntimeError('unexpected node: {!r}'.format(node))
       return (name, annotation)
 
+    assert node is not None
     name, annotation = parse_annotated_name(node)
     assert name not in '/*', repr(node)
 

--- a/docspec-python/src/docspec_python/parser.py
+++ b/docspec-python/src/docspec_python/parser.py
@@ -29,6 +29,8 @@ import re
 import textwrap
 import typing as t
 
+from nr.util.scanner import Scanner
+
 from docspec import (
   ApiObject,
   Argument,
@@ -269,7 +271,7 @@ class Parser:
       return_type=return_,
       decorations=decorations)
 
-  def parse_argument(self, node: t.Union[Leaf, Node], argtype: Argument.Type, scanner: 'ListScanner') -> Argument:
+  def parse_argument(self, node: t.Union[Leaf, Node], argtype: Argument.Type, scanner: 'Scanner[Leaf | Node]') -> Argument:
     """
     Parses an argument from the AST. *node* must be the current node at
     the current position of the *scanner*. The scanner is used to extract
@@ -278,11 +280,11 @@ class Parser:
 
     def parse_annotated_name(node):
       if node.type == syms.tname:
-        scanner = ListScanner(node.children)
+        scanner = Scanner(node.children)
         name = scanner.current.value
-        node = scanner.advance()
+        node = scanner.next()
         assert node.type == token.COLON, node.parent
-        node = scanner.advance()
+        node = scanner.next()
         annotation = self.nodes_to_string([node])
       elif node:
         name = node.value
@@ -292,6 +294,7 @@ class Parser:
       return (name, annotation)
 
     name, annotation = parse_annotated_name(node)
+    assert name not in '/*', repr(node)
 
     node = scanner.advance()
     default = None
@@ -304,7 +307,7 @@ class Parser:
 
   def parse_parameters(self, parameters):
     assert parameters.type == syms.parameters, parameters.type
-    result = []
+    result: t.List[Argument] = []
 
     arglist = find(lambda x: x.type == syms.typedargslist, parameters.children)
     if not arglist:
@@ -312,7 +315,7 @@ class Parser:
       #   not get wrapped in a `typedargslist`, but in a single `tname`.
       tname = find(lambda x: x.type == syms.tname, parameters.children)
       if tname:
-        scanner = ListScanner(parameters.children, parameters.children.index(tname))
+        scanner = Scanner(parameters.children, parameters.children.index(tname))
         result.append(self.parse_argument(tname, Argument.Type.POSITIONAL, scanner))
       else:
         # This must be either ["(", ")"] or ["(", "argname", ")"].
@@ -323,22 +326,34 @@ class Parser:
 
     argtype = Argument.Type.POSITIONAL
 
-    index = ListScanner(arglist.children)
-    for node in index.safe_iter(auto_advance=False):
+    index = Scanner(arglist.children)
+    for node in index.safe_iter():
       node = index.current
-      if node.type == token.STAR:
-        node = index.advance()
-        if node.type != token.COMMA:
+
+      if node.type == token.SLASH:
+        assert argtype == Argument.Type.POSITIONAL
+        # We need to retrospectively change the argument type of previous parsed arguments to POSITIONAL_ONLY.
+        for arg in result:
+          assert arg.type == Argument.Type.POSITIONAL, arg
+          arg.type = Argument.Type.POSITIONAL_ONLY
+        node = index.next()
+        if node.type == token.COMMA:
+          index.advance()
+
+      elif node.type == token.STAR:
+        node = index.next()
+        if node and node.type != token.COMMA:
           result.append(self.parse_argument(node, Argument.Type.POSITIONAL_REMAINDER, index))
         index.advance()
         argtype = Argument.Type.KEYWORD_ONLY
-        continue
-      elif node.type == token.DOUBLESTAR:
-        node = index.advance()
+
+      elif node and node.type == token.DOUBLESTAR:
+        node = index.next()
         result.append(self.parse_argument(node, Argument.Type.KEYWORD_REMAINDER, index))
-        continue
-      result.append(self.parse_argument(node, argtype, index))
-      index.advance()
+
+      else:
+        result.append(self.parse_argument(node, argtype, index))
+        index.advance()
 
     return result
 
@@ -360,9 +375,9 @@ class Parser:
   def parse_classdef_rawargs(self, classdef):
     metaclass = None
     bases = []
-    index = ListScanner(classdef.children, 2)
+    index = Scanner(classdef.children, 2)
     if index.current.type == token.LPAR:
-      index.advance()
+      index.next()
       while index.current.type != token.RPAR:
         if index.current.type == syms.argument:
           key = index.current.children[0].value
@@ -374,7 +389,7 @@ class Parser:
             pass
         else:
           bases.append(str(index.current))
-        index.advance()
+        index.next()
     return metaclass, bases
 
   def parse_classdef(self, parent, node, decorations):
@@ -550,80 +565,3 @@ class Parser:
       return ''.join(x.value for x in node.children)
     else:
       return node.value
-
-
-class ListScanner:
-  """
-  A helper class to navigate through a list. This is useful if you would
-  usually iterate over the list by index to be able to acces the next
-  element during the iteration.
-
-  Example:
-
-  ```py
-  scanner = ListScanner(lst)
-  for value in scanner.safe_iter():
-    if some_condition(value):
-      value = scanner.advance()
-  ```
-  """
-
-  def __init__(self, lst, index=0):
-    self._list = lst
-    self._index = index
-
-  def __bool__(self):
-    return self._index < len(self._list)
-
-  __nonzero__ = __bool__
-
-  @property
-  def current(self):
-    """
-    Returns the current list element.
-    """
-
-    return self._list[self._index]
-
-  def can_advance(self):
-    """
-    Returns `True` if there is a next element in the list.
-    """
-
-    return self._index < (len(self._list) - 1)
-
-  def advance(self, expect=False):
-    """
-    Advances the scanner to the next element in the list. If *expect* is set
-    to `True`, an #IndexError will be raised when there is no next element.
-    Otherwise, `None` will be returned.
-    """
-
-    self._index += 1
-    try:
-      return self.current
-    except IndexError:
-      if expect:
-        raise
-      return None
-
-  def safe_iter(self, auto_advance=True):
-    """
-    A useful generator function that iterates over every element in the list.
-    You may call #advance() during the iteration to retrieve the next
-    element in the list within a single iteration.
-
-    If *auto_advance* is `True` (default), the function generator will
-    always advance to the next element automatically. If it is set to `False`,
-    #advance() must be called manually in every iteration to ensure that
-    the scanner has advanced at least to the next element, or a
-    #RuntimeError will be raised.
-    """
-
-    index = self._index
-    while self:
-      yield self.current
-      if auto_advance:
-        self.advance()
-      elif self._index == index:
-        raise RuntimeError('next() has not been called on the ListScanner')

--- a/docspec-python/test/test_parser.py
+++ b/docspec-python/test/test_parser.py
@@ -24,10 +24,23 @@ from docspec_python import parse_python_module, ParserOptions
 from functools import wraps
 from io import StringIO
 from json import dumps
+from nr.util.inspect import get_callsite
 from textwrap import dedent
 from typing import List, Optional
 import pytest
 import sys
+
+
+def mkfunc(name: str, docstring: Optional[str], lineno: int, args: List[Argument]) -> Function:
+  return Function(
+    name=name,
+    location=None,
+    docstring=Docstring(docstring, Location(get_callsite().code_name, lineno)) if docstring else None,
+    modifiers=None,
+    args=args,
+    return_type=None,
+    decorations=[],
+  )
 
 
 def unset_location(obj: ApiObject):
@@ -183,22 +196,12 @@ def test_funcdef_5_single_stmt():
     return self.foo
   """
 
-  def mkfunc(name: str, docstring: Optional[str], lineno: int) -> Function:
-    return Function(
-      name=name,
-      location=None,
-      docstring=Docstring(docstring, Location('test_funcdef_5_single_stmt', lineno)) if docstring else None,
-      modifiers=None,
-      args=[Argument('self', Argument.Type.POSITIONAL, None, None, None)],
-      return_type=None,
-      decorations=[],
-    )
-
+  args = [Argument('self', Argument.Type.POSITIONAL, None, None, None)]
   return [
-    mkfunc('func1', None, 1),
-    mkfunc('func2', 'ABC\nDEF', 4),
-    mkfunc('func3', 'ABC\nDEF', 9),
-    mkfunc('func4', 'ABC\n  DEF', 14),
+    mkfunc('func1', None, 1, args),
+    mkfunc('func2', 'ABC\nDEF', 4, args),
+    mkfunc('func3', 'ABC\nDEF', 9, args),
+    mkfunc('func4', 'ABC\n  DEF', 14, args),
   ]
 
 
@@ -220,17 +223,6 @@ def test_funcdef_6_starred_args():
   def func5(abc, *, kwonly):
     '''Docstring goes here'''
   """
-
-  def mkfunc(name: str, docstring: Optional[str], lineno: int, args: List[Argument]) -> Function:
-    return Function(
-      name=name,
-      location=None,
-      docstring=Docstring(docstring, Location('test_funcdef_6_starred_args', lineno)) if docstring else None,
-      modifiers=None,
-      args=args,
-      return_type=None,
-      decorations=[],
-    )
 
   return [
     mkfunc('func1', None, 0, [
@@ -254,6 +246,36 @@ def test_funcdef_6_starred_args():
     ]),
   ]
 
+
+@pytest.mark.skipif(sys.version_info < (3, 8), reason="requires python3.8 or higher")
+@docspec_test()
+def test_funcdef_7_posonly_args():
+  """
+  def func1(x, y=3, /, z=5, w=7): pass
+  def func2(x, /, *v, a=1, b=2): pass
+  def func3(x, /, *, a=1, b=2, **kwargs): pass
+  """
+
+  return [
+    mkfunc('func1', None, 1, [
+      Argument('x', Argument.Type.POSITIONAL_ONLY),
+      Argument('y', Argument.Type.POSITIONAL_ONLY, default_value='3'),
+      Argument('z', Argument.Type.POSITIONAL, default_value='5'),
+      Argument('w', Argument.Type.POSITIONAL, default_value='7'),
+    ]),
+    mkfunc('func2', None, 2, [
+      Argument('x', Argument.Type.POSITIONAL_ONLY),
+      Argument('v', Argument.Type.POSITIONAL_REMAINDER),
+      Argument('a', Argument.Type.KEYWORD_ONLY, default_value='1'),
+      Argument('b', Argument.Type.KEYWORD_ONLY, default_value='2'),
+    ]),
+    mkfunc('func3', None, 3, [
+      Argument('x', Argument.Type.POSITIONAL_ONLY),
+      Argument('a', Argument.Type.KEYWORD_ONLY, default_value='1'),
+      Argument('b', Argument.Type.KEYWORD_ONLY, default_value='2'),
+      Argument('kwargs', Argument.Type.KEYWORD_REMAINDER),
+    ]),
+  ]
 
 @docspec_test()
 def test_classdef_1_exceptions():

--- a/docspec/pyproject.toml
+++ b/docspec/pyproject.toml
@@ -15,9 +15,11 @@ packages = [{include="docspec", from="src"}]
   Deprecated = "^1.2.12"
 
   [tool.poetry.dev-dependencies]
+  "databind.core" = "^1.2.1"
+  mypy = "*"
+  pytest = "*"
   types-Deprecated = "*"
   types-termcolor = "*"
-  "databind.core" = "^1.2.1"
 
   [tool.poetry.scripts]
   docspec = "docspec.__main__:main"

